lib/segment/src/index/field_index/full_text_index/inverted_index.rs

```diff
@@ -0,0 +1,193 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use serde::{Deserialize, Serialize};
+
+use crate::index::field_index::full_text_index::postings_iterator::intersect_btree_iterator;
+use crate::index::field_index::{CardinalityEstimation, PayloadBlockCondition, PrimaryCondition};
+use crate::types::{FieldCondition, Match, MatchText, PayloadKeyType, PointOffsetType};
+
+type PostingList = BTreeSet<PointOffsetType>;
+
+#[derive(Default, Serialize, Deserialize, Clone)]
+pub struct Document {
+    pub tokens: BTreeSet<String>,
+}
+
+impl Document {
+    pub fn is_empty(&self) -> bool {
+        self.tokens.is_empty()
+    }
+}
+
+pub struct ParsedQuery {
+    pub tokens: BTreeSet<String>,
+}
+
+impl ParsedQuery {
+    pub fn check_match(&self, document: &Document) -> bool {
+        // Check that all tokens are in document
+        self.tokens
+            .iter()
+            .all(|query_token| document.tokens.contains(query_token))
+    }
+}
+
+pub struct InvertedIndex {
+    postings: BTreeMap<String, PostingList>,
+    pub point_to_docs: Vec<Document>,
+    pub points_count: usize,
+}
+
+impl InvertedIndex {
+    pub fn new() -> InvertedIndex {
+        InvertedIndex {
+            postings: BTreeMap::new(),
+            point_to_docs: Vec::new(),
+            points_count: 0,
+        }
+    }
+
+    pub fn index_document(&mut self, idx: PointOffsetType, document: Document) {
+        for token in &document.tokens {
+            let posting = self
+                .postings
+                .entry(token.to_owned())
+                .or_insert_with(BTreeSet::new);
+            posting.insert(idx);
+        }
+        self.points_count += 1;
+        if self.point_to_docs.len() <= idx as usize {
+            self.point_to_docs
+                .resize(idx as usize + 1, Default::default());
+        }
+
+        self.point_to_docs[idx as usize] = document;
+    }
+
+    pub fn remove_document(&mut self, idx: PointOffsetType) -> Option<Document> {
+        if self.point_to_docs.len() <= idx as usize {
+            return None; // Already removed or never actually existed
+        }
+
+        let removed_doc = std::mem::take(&mut self.point_to_docs[idx as usize]);
+
+        if removed_doc.is_empty() {
+            return None;
+        }
+        self.points_count -= 1;
```

Comment:
I mentioned case of documents with empty token. Here is a bug. When we add an empty document and remove it, `self.points_count` is wrong